scalasql/query/src-3/TableMacro.scala

```diff
@@ -0,0 +1,177 @@
+package scalasql.query
+
+import scalasql.core.{DialectTypeMappers, Expr => SqlExpr, Queryable, Sc, TypeMapper}
+import scala.compiletime.summonInline
+import scala.quoted.*
+
+object TableMacros {
+  def applyImpl[V[_[_]] <: Product](using Quotes, Type[V]): Expr[Table.Metadata[V]] = {
+    import quotes.reflect.*
+
+    val caseClassType = TypeRepr.of[V]
+    val constructor = caseClassType.typeSymbol.primaryConstructor
+    val constructorTypeParams = constructor.paramSymss(0)
+    val constructorValueParams = constructor.paramSymss(1)
+
+    def paramType(param: Symbol): TypeRepr = caseClassType.memberType(param)
+
+    def isTypeParamType(param: Symbol): Boolean =
+      paramType(param).typeSymbol.toString != constructorTypeParams.head.toString
+
+    def subParam(paramInfo: TypeRepr, tpe: TypeRepr): TypeRepr =
+      paramInfo.substituteTypes(List(constructorTypeParams.head), List(tpe))
+
+    val queryables = '{ (dialect: DialectTypeMappers, n: Int) =>
+      {
+        @annotation.nowarn("msg=unused")
+        given d: DialectTypeMappers = dialect
```

Comment:
Do we actually need the name `d` here, or can we do `given DialectTypeMappers = dialect`?

Also, why do we need the `nowarn` annotation here, since I assume the `DialectTypeMappers` instance should be used by the following generated `summonInline` call?

Reply:
Nope, no need for the name, good call.

The compiler reports it as unused, I guess because in the context of the macro definition it doesn't know that it's used yet, it only knows once the result of the macro is expanded and typechecked...?

That said, I think this could trigger warnings in downstream users code if either of the following are true:

1. The class has no fields, or only has fields with `TypeMapper` instances defined outside of `DialectTypeMappers`, so this instance isn't necessary -- I confirmed that this is then reported as unused after macro expansion
2. The class has fields with `TypeMapper` instances defined within `DialectTypeMappers`, so this instance _is_ necessary -- in this case, if the user has the `-Wunused:nowarn` option enabled, then the compiler may report this `@annotation.nowarn` as one that doesn't suppress any warnings. I haven't been able to confirm this, but it's theoretically possible

To work around both of these, I updated this to ensure the `given` is always used, both in the macro definition and after macro expansion:

```scala
given DialectTypeMappers = dialect
lazy val _ = summon[DialectTypeMappers]
```